text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,299 @@
+  - **Example:** The page contains a footer that never changes.
+  - **Example:** The landing page is mostly static and is cached on a CDN.
+- **Exception C:** Render-critical state is changed between the time of SSR and CSR rehydration, causing an unexpected mismatch (Fig 1d).
+  - **Example:** A product price is updated after SSR but before rehydration on the client.
+  - **Example:** The CDN delivers a stale version of the landing page, which was updated after the SSR HTML was en route to the client.
+
+<figure>
+  <img src="../assets/0000-ssr-rehydration-fig1.png" alt="Fig. 1">
+  <figcaption align = "center">
+    <b>Fig.1: the happy path and its exceptions</b>
+  </figcaption>
+</figure>
+
+## Design
+
+Several pieces need to come together to unlock rehydration. The general flow is illustrated below in figure 2.
+
+<figure>
+  <img src="../assets/0000-ssr-rehydration-fig2.png" alt="Fig. 2" width="60%">
+  <figcaption align = "center">
+    <b>Fig.2: the proposed implementation</b>
+  </figcaption>
+</figure><br /><br />
+
+In roughly chronological order with respect to a single page request, the following steps shall be taken as part of rehydration:
+
+- As part of SSR, serialize render-critical state and bake it into the HTML response returned from the server. For each `LightningElement`, attach corresponding state to its root DOM node as `data-ssr-state`.
+
+- Rather than calling `LWC#createElement`, required that `LWC#hydrateElement` is called to mount the root Lightning Web Component on the client.
+
+- Generate server-side VDOM from the SSR HTML on the client.
+
+- As the initial client-side render progresses, extract state from `data-ssr-state` attributes, instantiate `LightningElements` with extracted state, and generate client-side VDOM.
+
+- Patch server-side VDOM with client-side VDOM.
+  
+  - Where `data-ssr-preserve` is detected, extract the server-side VDOM subtree and store for subsequent client-side renders.
+  
+  - Attach event listeners and non-serializable props.
+  
+  - Resolve unexpected inconsistencies in favor of the client-side VDOM, addressing [Exception C to the happy path](#the-happy-path). With the exception of placeholder components, this case should be rare, since the first CSR will rely on state baked into SSR as `data-ssr-state`.
+
+### `hydrateElement`
+
+In a typical LWC web application, a developer might mount their app using the following pattern:
+
+```javascript
+import { createElement } from 'lwc';
+import MyApp from './my-app';
+
+document
+  .querySelector("#root")
+  .appendChild(createElement('my-app', { is: MyApp }));
+```
+
+After the proposed change, if a developer wants to take advantage of rehydration, a developer would instead do:
+
+```javascript
+import { hydrateElement } from 'lwc';
+import MyApp from './my-app';
+
+hydrateElement(MyApp, document.querySelector('#root')); 
+```
+
+### Server-only Component Support

```

Comment:
It feels out of the scope for this first iteration of rehydration. What do you think about moving this section to a different proposal?

Reply:
I am not opposed to that. That is to say, we will definitely not require that for any MVP.

Reply:
I'm happy to push this to a different proposal.